File: backend/schemas/ticket.py
"""
工单相关 Pydantic Schemas
"""
from datetime import datetime
from decimal import Decimal
from typing import Optional, Literal
from pydantic import BaseModel, Field

from .common import PageParams


# 工单类型
TicketTypeLiteral = Literal["membership", "recharge"]

# 工单状态（预留多种便于扩展）
TicketStatusLiteral = Literal["pending", "processing", "completed", "rejected", "failed"]

# 会员周期
PeriodTypeLiteral = Literal["monthly", "quarterly", "yearly"]


class TicketMembershipCreate(BaseModel):
    """开通会员工单创建参数"""
    level_code: str = Field(..., description="等级代码：vip/svip/max")
    vip_expire_date: str = Field(..., description="VIP到期时间，格式：YYYY-MM-DD")
    is_paid: bool = Field(..., description="是否已收费")
    payment_method: Optional[str] = Field(None, description="收费方式：微信/支付宝/对公转账等")
    voucher: Optional[str] = Field(None, description="凭证（图片URL或文字说明）")
    period_type: PeriodTypeLiteral = Field(..., description="会员周期：monthly-月度, quarterly-季度, yearly-年度")


class TicketRechargeCreate(BaseModel):
    """充值算力工单创建参数"""
    amount: Decimal = Field(..., gt=0, description="充值金额（算力点）")


class TicketCreate(BaseModel):
    """工单创建请求"""
    type: TicketTypeLiteral = Field(..., description="工单类型：membership-开通会员, recharge-充值算力")
    user_id: int = Field(..., description="目标用户ID")
    # 会员工单必填字段
    membership: Optional[TicketMembershipCreate] = Field(None, description="会员工单详情（type=membership时必填）")
    # 充值工单必填字段
    recharge: Optional[TicketRechargeCreate] = Field(None, description="充值工单详情（type=recharge时必填）")
    remark: Optional[str] = Field(None, description="备注说明")


class TicketQueryParams(PageParams):
    """工单查询参数"""
    type: Optional[TicketTypeLiteral] = Field(None, description="工单类型")
    status: Optional[TicketStatusLiteral] = Field(None, description="工单状态")
    user_id: Optional[int] = Field(None, description="目标用户ID")
    creator_id: Optional[int] = Field(None, description="创建人ID")


class TicketUserInfo(BaseModel):
    """工单关联用户简要信息"""
    id: str = Field(..., description="用户ID")
    username: str = Field(..., description="用户名")
    phone: Optional[str] = Field(None, description="手机号")
    nickname: Optional[str] = Field(None, description="昵称")


class TicketCreatorInfo(BaseModel):
    """工单创建人简要信息"""
    id: int = Field(..., description="管理员ID")
    username: str = Field(..., description="管理员用户名")


class TicketResponse(BaseModel):
    """工单响应"""
    id: int = Field(..., description="工单ID")
    type: str = Field(..., description="工单类型")
    status: str = Field(..., description="工单状态")
    user_id: int = Field(..., description="目标用户ID")
    user: Optional[TicketUserInfo] = Field(None, description="目标用户信息")
    creator_id: int = Field(..., description="创建人ID")
    creator: Optional[TicketCreatorInfo] = Field(None, description="创建人信息")
    handler_id: Optional[int] = Field(None, description="处理人ID")
    handler: Optional[TicketCreatorInfo] = Field(None, description="处理人信息")
    # 会员工单字段
    is_paid: Optional[bool] = Field(None, description="是否已收费")
    payment_method: Optional[str] = Field(None, description="收费方式")
    voucher: Optional[str] = Field(None, description="凭证")
    period_type: Optional[str] = Field(None, description="会员周期")
    # 扩展数据（解析后）
    extra_data: Optional[dict] = Field(None, description="扩展数据：level_code/vip_expire_date 或 amount")
    remark: Optional[str] = Field(None, description="备注")
    handled_at: Optional[datetime] = Field(None, description="处理时间")
    fail_reason: Optional[str] = Field(None, description="处理失败原因")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: Optional[datetime] = Field(None, description="更新时间")

    class Config:
        from_attributes = True

File: frontend/src/api/modules/ticket.ts
import { ResPage } from "@/api/interface/index";
import { PORT1 } from "@/api/config/servicePort";
import http from "@/api";

/**
 * @name 工单管理模块
 * 对接 /api/v1/admin/tickets 接口
 */

/** 工单类型 */
export type TicketType = "membership" | "recharge";

/** 工单状态 */
export type TicketStatus = "pending" | "processing" | "completed" | "rejected" | "failed";

/** 会员周期 */
export type PeriodType = "monthly" | "quarterly" | "yearly";

/** 工单列表项 */
export interface TicketItem {
  id: number;
  type: TicketType;
  status: TicketStatus;
  user_id: number;
  user?: { id: string; username: string; phone?: string; nickname?: string };
  creator_id: number;
  creator?: { id: number; username: string };
  handler_id?: number;
  handler?: { id: number; username: string };
  is_paid?: boolean;
  payment_method?: string;
  voucher?: string;
  period_type?: string;
  extra_data?: Record<string, unknown>;
  remark?: string;
  handled_at?: string;
  fail_reason?: string;
  created_at: string;
  updated_at?: string;
}

/** 工单查询参数 */
export interface TicketQueryParams {
  pageNum: number;
  pageSize: number;
  type?: TicketType;
  status?: TicketStatus;
  user_id?: number;
  creator_id?: number;
}

/** 创建会员工单参数 */
export interface TicketMembershipCreate {
  level_code: string;
  vip_expire_date: string;
  is_paid: boolean;
  payment_method?: string;
  voucher?: string;
  period_type: PeriodType;
}

/** 创建充值工单参数 */
export interface TicketRechargeCreate {
  amount: number;
}

/** 创建工单参数 */
export interface TicketCreateParams {
  type: TicketType;
  user_id: number;
  membership?: TicketMembershipCreate;
  recharge?: TicketRechargeCreate;
  remark?: string;
}

// 获取工单列表
export const getTicketList = (params: TicketQueryParams) => {
  return http.get<ResPage<TicketItem>>(PORT1 + `/tickets`, params);
};

// 获取工单详情
export const getTicketDetail = (id: number) => {
  return http.get<TicketItem>(PORT1 + `/tickets/${id}`);
};

// 创建工单
export const createTicket = (params: TicketCreateParams) => {
  return http.post<{ id: number }>(PORT1 + `/tickets`, params);
};

// 处理工单
export const handleTicket = (id: number) => {
  return http.post<TicketItem>(PORT1 + `/tickets/${id}/handle`);
};
